bpf/process/bpf_execve_bprm_commit_creds.c

```diff
@@ -76,12 +79,60 @@ BPF_KPROBE(tg_kp_bprm_committing_creds, struct linux_binprm *bprm)
 	gid = BPF_CORE_READ(task, cred, gid.val);
 
 	/* Is setuid? */
-	if (euid != uid)
+	if (euid != uid) {
 		heap->info.secureexec |= EXEC_SETUID;
+		/* If euid is being changed to root? */
+		ruid = BPF_CORE_READ(bprm, cred, uid.val);
+		if (!__is_uid_global_root(ruid) && __is_uid_global_root(euid))
```

Comment:
What is `ruid` and why we check whether it's not zero? Is there a case where `uid != 0` but `ruid == 0`?
I guess it's whatever the kernel will call `commit_creds` with?

Reply:
1. ruid is the real user id that started the command so we check if it was not global root (this won't change for setuid execution it will keep its value)
2. Yes you could start with ruid == 0 as real root and execute a setuid binary to 1000 uid as an example, but for this case we will only report the setuid=1000 but we won't report the privilege `PRIVILEGES_RAISED_EXEC_FILE_SETUID` as it is more of a privilege drop, and such case usually don't happen in practice. It is more privilege raise to root. So that check is to ensure this logic.
3. For the commit_creds yes it happens later, we are in middle after the binary was parsed and its inode owners were read by kernel, then the capability LSM did some magic fixup to ensure everything is right, then our program is triggered just before the final commit_creds on the task cred, so we operate on the bprm (binary prameters) creds , and there won't be any change anymore what we are parsing is what will be committed in commit_creds

Reply:
> Yes you could start with ruid == 0 as real root and execute a setuid binary to 1000 uid as an example, 

What  would be the `ruid` and `uid` values in this case?